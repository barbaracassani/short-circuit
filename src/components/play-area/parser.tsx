import React from "react";

import config from '../../config';

import Level from "../../interfaces/level";
import Tile from "../../components/tile/Tile";

interface Sets {
  [index: string]: string;
}

const sets: Sets = {
  UmbrellaStarSet: "http://localhost:3000/sets/umbrella-star.svg"
};

const iterateGrid = (levelGrid: Level["grid"]): (JSX.Element|null)[] => {
  const url = sets.UmbrellaStarSet;
  const identifier = "umbrella";

  const splitString = levelGrid.split("");
  const tileSize = config.tileSize;
  let y = 0;
  let x = -tileSize;
  return splitString.map((char: string, index: number) => {
    x += tileSize;
    if (char === "\n") {
      y += tileSize;
      x = -tileSize;
      return null;
    } else {
        return (
            <Tile
                xPos={x}
                yPos={y}
                key={index}
                getCurrentState={() => {}}
                identifier={identifier}
            >
                <svg className={`icon icon-${identifier}`}>
                    <use xlinkHref={`${url}#icon-${identifier}`} />
                </svg>
            </Tile>
        );
    }

  });
};

const parser = (level: Level) => {
  const url = sets.UmbrellaStarSet;
  const identifier = "umbrella";
  return iterateGrid(level.grid);
};

export default parser;
