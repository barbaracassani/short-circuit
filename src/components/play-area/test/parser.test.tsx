import React from "react";
import ReactDOM from "react-dom";
import parser from "../parser";
import Level from "../../../interfaces/level";
import { stripIndents } from "common-tags";

jest.mock("../../tile/Tile", () => {
  return {
    __esModule: true,
    default: (props: any) => {
      return <div {...props}></div>;
    }
  };
});

describe("the output of parser", () => {
  it("returns one tile", () => {
    const oneTile = `x`;
    expect(
      parser({
        grid: oneTile
      } as Level)
    ).toMatchSnapshot("one tile");
  });
  it("returns many tiles", () => {
    const manyTiles = stripIndents`xxx
                                   xxx`;
    expect(
      parser({
        grid: manyTiles
      } as Level)
    ).toMatchSnapshot("many tiles");
  });
});
