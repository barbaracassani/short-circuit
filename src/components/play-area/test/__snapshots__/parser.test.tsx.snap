// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the output of parser returns many tiles: many tiles 1`] = `
Array [
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={0}
    yPos={0}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={65}
    yPos={0}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={130}
    yPos={0}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
  null,
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={0}
    yPos={65}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={65}
    yPos={65}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={130}
    yPos={65}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
]
`;

exports[`the output of parser returns one tile: one tile 1`] = `
Array [
  <default
    getCurrentState={[Function]}
    identifier="umbrella"
    xPos={0}
    yPos={0}
  >
    <svg
      className="icon icon-umbrella"
    >
      <use
        xlinkHref="http://localhost:3000/sets/umbrella-star.svg#icon-umbrella"
      />
    </svg>
  </default>,
]
`;
